File: view-mode/src/androidTest/java/com/jaynewstrom/viewmode/ViewModeViewTest.java
package com.jaynewstrom.viewmode;

import android.annotation.SuppressLint;
import android.content.Context;
import android.support.test.InstrumentationRegistry;
import android.support.test.annotation.UiThreadTest;
import android.support.test.runner.AndroidJUnit4;
import android.view.ContextThemeWrapper;
import android.view.View;

import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(AndroidJUnit4.class)
public final class ViewModeViewTest {

    private Context context;

    @SuppressLint("PrivateResource")
    @Before public void setupContext() {
        // CoordinatorLayout needs a context that extends from Base_Theme_AppCompat
        context = new ContextThemeWrapper(InstrumentationRegistry.getTargetContext(),
                android.support.design.R.style.Base_Theme_AppCompat);
    }

    @Test @UiThreadTest public void ensureCurrentViewModeIsNullByDefault() {
        ViewModeView view = new ViewModeView(context);
        Assert.assertNull(view.currentViewMode());
    }

    @Test @UiThreadTest public void whenCachingIsEnabledEnsureShowingTheViewCachesTheView() {
        ViewModeView view = new ViewModeView(context);
        ViewMode viewMode = mock(ViewMode.class);
        View cachedView = new View(view.getContext());
        when(viewMode.createView(view)).thenReturn(cachedView, new View(view.getContext()));
        view.showViewMode(viewMode);
        Assert.assertEquals(cachedView, view.viewForViewMode(viewMode));
    }

    @Test @UiThreadTest public void whenCachingIsEnabledEnsureTheSameViewIsReturned() {
        ViewModeView view = new ViewModeView(context);
        ViewMode viewMode = mock(ViewMode.class);
        when(viewMode.createView(view)).thenReturn(new View(view.getContext()), new View(view.getContext()));
        Assert.assertEquals(view.viewForViewMode(viewMode), view.viewForViewMode(viewMode));
    }

    @Test @UiThreadTest public void whenCachingIsDisabledEnsureDifferentViewsAreReturned() {
        ViewModeView view = new ViewModeView(context);
        view.setCacheViews(false);
        ViewMode viewMode = mock(ViewMode.class);
        when(viewMode.createView(view)).thenReturn(new View(view.getContext()), new View(view.getContext()));
        Assert.assertNotEquals(view.viewForViewMode(viewMode), view.viewForViewMode(viewMode));
    }

    @Test @UiThreadTest public void ensureChangingViewModesOnlyCreatesTheViewOnce() {
        ViewModeView view = new ViewModeView(context);
        ViewMode viewMode = mock(ViewMode.class);
        when(viewMode.createView(view)).thenReturn(new View(view.getContext()), new View(view.getContext()));
        view.showViewMode(viewMode);
        view.showViewMode(viewMode);
        verify(viewMode, times(1)).createView(view);
    }

    @Test @UiThreadTest public void ensureOnlyOneChildViewIsVisible() {
        ViewModeView view = new ViewModeView(context);
        ViewMode viewModeOne = mock(ViewMode.class);
        when(viewModeOne.createView(view)).thenReturn(new View(view.getContext()));
        ViewMode viewModeTwo = mock(ViewMode.class);
        when(viewModeTwo.createView(view)).thenReturn(new View(view.getContext()));
        view.showViewMode(viewModeOne);
        view.showViewMode(viewModeTwo);
        Assert.assertEquals(2, view.getChildCount());
        int visibleChildren = 0;
        for (int i = 0, childCount = view.getChildCount(); i < childCount; i++) {
            if (view.getChildAt(i).getVisibility() == View.VISIBLE) {
                visibleChildren++;
            }
        }
        Assert.assertEquals(1, visibleChildren);
    }
}

File: view-mode/src/androidTest/java/com/jaynewstrom/viewmode/ViewModeViewTest.kt
package com.jaynewstrom.viewmode

import android.annotation.SuppressLint
import android.content.Context
import android.view.ContextThemeWrapper
import android.view.View
import androidx.test.annotation.UiThreadTest
import androidx.test.core.app.ApplicationProvider
import androidx.test.ext.junit.runners.AndroidJUnit4
import org.junit.Assert
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mockito.`when`
import org.mockito.Mockito.mock
import org.mockito.Mockito.times
import org.mockito.Mockito.verify
import java.util.concurrent.CountDownLatch

@RunWith(AndroidJUnit4::class)
class ViewModeViewTest {

    private lateinit var context: Context

    @SuppressLint("PrivateResource")
    @Before
    fun setupContext() {
        // CoordinatorLayout needs a context that extends from Base_Theme_AppCompat
        context = ContextThemeWrapper(
            ApplicationProvider.getApplicationContext<Context>(),
            androidx.appcompat.R.style.Base_Theme_AppCompat
        )
    }

    @Test @UiThreadTest fun ensureCurrentViewModeIsNullByDefault() {
        val view = ViewModeView(context)
        Assert.assertNull(view.currentViewMode())
    }

    @Test @UiThreadTest fun whenCachingIsEnabledEnsureShowingTheViewCachesTheView() {
        val view = ViewModeView(context)
        val viewMode = mock(ViewMode::class.java)
        val cachedView = View(view.context)
        `when`(viewMode.createView(view)).thenReturn(cachedView, View(view.context))
        view.showViewMode(viewMode)
        Assert.assertEquals(cachedView, view.viewForViewMode(viewMode))
    }

    @Test @UiThreadTest fun whenCachingIsEnabledEnsureTheSameViewIsReturned() {
        val view = ViewModeView(context)
        val viewMode = mock(ViewMode::class.java)
        `when`(viewMode.createView(view)).thenReturn(View(view.context), View(view.context))
        Assert.assertEquals(view.viewForViewMode(viewMode), view.viewForViewMode(viewMode))
    }

    @Test @UiThreadTest fun whenCachingIsDisabledEnsureDifferentViewsAreReturned() {
        val view = ViewModeView(context)
        view.setCacheViews(false)
        val viewMode = mock(ViewMode::class.java)
        `when`(viewMode.createView(view)).thenReturn(View(view.context), View(view.context))
        Assert.assertNotEquals(view.viewForViewMode(viewMode), view.viewForViewMode(viewMode))
    }

    @Test @UiThreadTest fun ensureChangingViewModesOnlyCreatesTheViewOnce() {
        val view = ViewModeView(context)
        val viewMode = mock(ViewMode::class.java)
        `when`(viewMode.createView(view)).thenReturn(View(view.context), View(view.context))
        view.showViewMode(viewMode)
        view.showViewMode(viewMode)
        verify(viewMode, times(1)).createView(view)
    }

    @Test @UiThreadTest fun ensureOnlyOneChildViewIsVisible() {
        val view = ViewModeView(context)
        val viewModeOne = mock(ViewMode::class.java)
        `when`(viewModeOne.createView(view)).thenReturn(View(view.context))
        val viewModeTwo = mock(ViewMode::class.java)
        `when`(viewModeTwo.createView(view)).thenReturn(View(view.context))
        view.showViewMode(viewModeOne)
        view.showViewMode(viewModeTwo)
        Assert.assertEquals(2, view.childCount.toLong())
        var visibleChildren = 0
        var i = 0
        val childCount = view.childCount
        while (i < childCount) {
            if (view.getChildAt(i).visibility == View.VISIBLE) {
                visibleChildren++
            }
            i++
        }
        Assert.assertEquals(1, visibleChildren.toLong())
    }

    @Test @UiThreadTest fun whenCachingIsDisabledEnsureOnlyOneChildIsAddedToTheView() {
        val view = ViewModeView(context)
        view.setCacheViews(false)
        val viewModeOne = mock(ViewMode::class.java)
        `when`(viewModeOne.createView(view)).thenReturn(View(view.context))
        val viewModeTwo = mock(ViewMode::class.java)
        `when`(viewModeTwo.createView(view)).thenReturn(View(view.context))
        view.showViewMode(viewModeOne)
        view.showViewMode(viewModeTwo)
        Assert.assertEquals(1, view.childCount.toLong())
    }

    @Test @UiThreadTest fun transitionFailsWhenAnotherTransitionIsInProgress() {
        val view = ViewModeView(context)
        val viewModeOne = mock(ViewMode::class.java)
        `when`(viewModeOne.createView(view)).thenReturn(View(view.context))
        val viewModeTwo = mock(ViewMode::class.java)
        `when`(viewModeTwo.createView(view)).thenReturn(View(view.context))
        view.showViewMode(viewModeOne, object : ViewModeTransition {
            override fun transition(fromViewMode: ViewMode?, fromView: View?, toView: View, transitionComplete: () -> Unit) {
                // Not calling transitionComplete
            }
        })
        assertThrowsWithMessage<IllegalStateException>("A transition is already occurring.") {
            view.showViewMode(viewModeTwo)
        }
    }

    @Test @UiThreadTest fun transitionSendsCorrectValues() {
        val view = ViewModeView(context)
        val viewModeOne = mock(ViewMode::class.java)
        val viewOne = View(view.context)
        `when`(viewModeOne.createView(view)).thenReturn(viewOne)
        val viewModeTwo = mock(ViewMode::class.java)
        val viewTwo = View(view.context)
        `when`(viewModeTwo.createView(view)).thenReturn(viewTwo)
        val countDownLatch = CountDownLatch(2)
        view.showViewMode(viewModeOne, object : ViewModeTransition {
            override fun transition(fromViewMode: ViewMode?, fromView: View?, toView: View, transitionComplete: () -> Unit) {
                assertNull(fromViewMode)
                assertNull(fromView)
                assertEquals(toView, viewOne)
                transitionComplete()
                countDownLatch.countDown()
            }
        })
        view.showViewMode(viewModeTwo, object : ViewModeTransition {
            override fun transition(fromViewMode: ViewMode?, fromView: View?, toView: View, transitionComplete: () -> Unit) {
                assertEquals(fromViewMode, viewModeOne)
                assertEquals(fromView, viewOne)
                assertEquals(toView, viewTwo)
                transitionComplete()
                countDownLatch.countDown()
            }
        })
        countDownLatch.await()
    }
}
